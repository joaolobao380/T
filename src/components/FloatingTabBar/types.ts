import { BarChartHorizontalBig, Home, Menu } from 'lucide-react-native';

export type IconComponentsType = {
  Home: typeof Home;
  Report: typeof BarChartHorizontalBig;
  Menu: typeof Menu;
};
