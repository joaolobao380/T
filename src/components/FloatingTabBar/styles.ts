import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    backgroundColor: 'transparent',
    borderRadius: 100,
    bottom: 20,
    marginHorizontal: 36,
    width: 167,
    height: 58,
    alignSelf: 'center',
    paddingHorizontal: 1,
    position: 'absolute',
  },
  tabButton: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  gradientBackground: {
    borderRadius: 100,
    height: '100%',
    width: '100%',
    flexDirection: 'row',
    paddingHorizontal: 1,
  },
  containerIconComponentFocused: {
    backgroundColor: 'white',
    padding: 15,
    borderRadius: 27,
    width: 54,
    height: 54,
    alignItems: 'center',
    justifyContent: 'center',
  },
  containerIconComponenteNotFocused: {
    backgroundColor: '#2A2A2A',
    padding: 0,
    borderRadius: 0,
    width: 0,
    height: 0,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
