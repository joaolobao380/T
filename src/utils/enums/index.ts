export enum messageToast {
  ERROR = 'customError',
  SUCCESSFUL = 'successful',
}

export enum errorMessageFirebase {
  TOO_MANY = 'auth/too-many-requests',
  INVALID_PASSWORD = 'auth/invalid-password',
  INVALID_EMAIL = 'auth/invalid-email',
  INVALID_CREDENTIAL = 'auth/invalid-credential',
}

export enum textToast {
  ATTENTION = 'Atenção',
  SUCCESSFUL = 'Sucesso!',
}

export enum positionToast {
  POSITION = 'bottom',
}

export enum keyAsyncUser {
  USER_TOKEN_KEY = '@userToken',
}

export enum navBottomName {
  HOME = 'Home',
  REPORT = 'Report',
  MENU = 'Menu',
}
