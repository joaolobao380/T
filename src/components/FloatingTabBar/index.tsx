import { LinearGradient } from 'expo-linear-gradient';
import { BarChartHorizontalBig, Home, Menu } from 'lucide-react-native';
import React from 'react';
import { View, TouchableOpacity } from 'react-native';

import { styles } from './styles';
import { IconComponentsType } from './types';

const iconComponents: IconComponentsType = {
  Home,
  Report: BarChartHorizontalBig,
  Menu,
};

const FloatingTabBar: React.FC<any> = ({ state, descriptors, navigation }) => {
  return (
    <View style={styles.container}>
      <LinearGradient
        colors={['rgba(46, 46, 46, 0.8)', '#111111', '#0A0A0A']}
        start={{ x: 0.5, y: 0 }}
        end={{ x: 0.5, y: 1 }}
        style={styles.gradientBackground}>
        {state.routes.map((route: any, index: number) => {
          const { options } = descriptors[route.key];
          const isFocused = state.index === index;

          const onPress = () => {
            const event = navigation.emit({
              type: 'tabPress',
              target: route.key,
              canPreventDefault: true,
            });

            if (!isFocused && !event.defaultPrevented) {
              navigation.navigate(route.name);
            }
          };

          const IconComponent = iconComponents[route.name as keyof IconComponentsType];

          return (
            <TouchableOpacity
              key={index}
              accessibilityRole="button"
              accessibilityState={isFocused ? { selected: true } : {}}
              accessibilityLabel={options.tabBarAccessibilityLabel}
              testID={options.tabBarTestID}
              onPress={onPress}
              style={styles.tabButton}>
              <View
                style={
                  isFocused
                    ? styles.containerIconComponentFocused
                    : styles.containerIconComponenteNotFocused
                }>
                <IconComponent color={isFocused ? '#2A2A2A' : 'white'} size={25} />
              </View>
            </TouchableOpacity>
          );
        })}
      </LinearGradient>
    </View>
  );
};

export default FloatingTabBar;
