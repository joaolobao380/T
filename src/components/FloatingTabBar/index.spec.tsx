import { render, fireEvent } from '@testing-library/react-native';
import { navBottomName } from '@utils/enums';
import React from 'react';

import FloatingTabBar from '.';

jest.mock('expo-linear-gradient', () => ({
  LinearGradient: ({ children }: any) => children,
}));

jest.mock('lucide-react-native', () => ({
  Home: 'HomeIcon',
  BarChartHorizontalBig: 'BarChartIcon',
  Menu: 'MenuIcon',
}));

describe('FloatingTabBar', () => {
  const mockNavigation = {
    navigate: jest.fn(),
    emit: jest.fn(() => ({ defaultPrevented: false })),
  } as any;

  const mockState = {
    index: 0,
    routes: [
      { key: navBottomName.HOME, name: navBottomName.HOME },
      { key: navBottomName.REPORT, name: navBottomName.REPORT },
      { key: navBottomName.MENU, name: navBottomName.MENU },
    ],
  } as any;

  const mockDescriptors = {
    Home: { options: { tabBarTestID: 'Home-icon' } },
    Report: { options: { tabBarTestID: 'Report-icon' } },
    Menu: { options: { tabBarTestID: 'Menu-icon' } },
  } as any;

  it('renders all icons', () => {
    const { getByTestId } = render(
      <FloatingTabBar state={mockState} descriptors={mockDescriptors} navigation={mockNavigation} />
    );

    expect(getByTestId('Home-icon')).toBeTruthy();
    expect(getByTestId('Report-icon')).toBeTruthy();
    expect(getByTestId('Menu-icon')).toBeTruthy();
  });

  it('navigates to the correct screen when an icon is pressed', () => {
    const { getByTestId } = render(
      <FloatingTabBar state={mockState} descriptors={mockDescriptors} navigation={mockNavigation} />
    );

    fireEvent.press(getByTestId('Report-icon'));
    expect(mockNavigation.navigate).toHaveBeenCalledWith(navBottomName.REPORT);
  });

  it('renders correctly', () => {
    const { toJSON } = render(
      <FloatingTabBar state={mockState} descriptors={mockDescriptors} navigation={mockNavigation} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
