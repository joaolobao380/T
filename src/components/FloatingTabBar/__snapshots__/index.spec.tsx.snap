// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingTabBar renders correctly 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "backgroundColor": "transparent",
      "borderRadius": 100,
      "bottom": 20,
      "flexDirection": "row",
      "height": 58,
      "marginHorizontal": 36,
      "paddingHorizontal": 1,
      "position": "absolute",
      "width": 167,
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": true,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID="Home-icon"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 27,
          "height": 54,
          "justifyContent": "center",
          "padding": 15,
          "width": 54,
        }
      }
    >
      <HomeIcon
        color="#2A2A2A"
        size={25}
      />
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID="Report-icon"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#2A2A2A",
          "borderRadius": 0,
          "height": 0,
          "justifyContent": "center",
          "padding": 0,
          "width": 0,
        }
      }
    >
      <BarChartIcon
        color="white"
        size={25}
      />
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID="Menu-icon"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#2A2A2A",
          "borderRadius": 0,
          "height": 0,
          "justifyContent": "center",
          "padding": 0,
          "width": 0,
        }
      }
    >
      <MenuIcon
        color="white"
        size={25}
      />
    </View>
  </View>
</View>
`;
